### PR TITLE
Refactor OMs to use "m" property of Pokemon

### PR DESCRIPTION
**Pull request by mathfreak231**
Quite often temporary OMs wind up needing to store arbitrary properties to a Pokemon, which TypeScript won't allow. The solution, brought up on #4952 , was to give Pokemon a "data" property with type `AnyObject`. This reduces the need for ts-ignores and properties that were only defined to be used by one OM.

~~"data" is a working name; other possibilities include clearer names like "miscData" or "otherData", or single-letter names like "d"~~ The name "m" was decided

### CODE REVIEW

**Comment by Zarel** [comment]
``.m`` for "mod data" might be nice. We do already use ``Dex.data`` to store the Dex data tables, so a different name here might be important.

**Comment by Zarel** [comment]
Looks good! I like it a lot!

**Comment by Zarel** [comment]
Do you need help rebasing?

**Comment by mathfreak231** [comment]
probably 😔

**Comment by Zarel** [comment]
Use these bash commands (in Windows, type them into Git Bash).

```
git remote add upstream git@github.com:Zarel/Pokemon-Showdown.git
```

This lets you access the official repository using the name `upstream`.

```
git checkout master
git pull upstream master
git push
```

This updates your master branch to the latest version from upstream.

```
git checkout pokemon-dot-data
git rebase master
```

This takes your commits from `pokemon-dot-data`, and puts them on top of the latest master.

At this point, you'll get a conflict in `config/formats.js` Go and edit that file and handle the merge. VS Code has some merge tools, but any text editor should make it reasonably clear.

```
git add .
git rebase --continue
```

You might get multiple conflicts, in which case you'll have to edit the conflicting file, and then run `git add .` and `git rebase --continue` again.

At some point, Git will be done rebasing.

```
git push --force-with-lease
```

This will push your rebased changes. The `--force` part is necessary because rebasing changes history, which Git considers a somewhat-dangerous operation.

**Comment by Zarel** [comment]
Yay! Looks like you were successful in rebasing.

File: mods/ssb/statuses.js [hunk 1]
```diff
@@ -1135,29 +1132,24 @@ let BattleStatuses = {
...
 			// Now we handle the fake claim "fainting"
-			// @ts-ignore Hack for Snaquaza's Z move
-			pokemon.hp = pokemon.claimHP;
+			pokemon.hp = pokemon.m.claimHP;
 			pokemon.formeChange(pokemon.baseTemplate.id);
 			pokemon.moveSlots = pokemon.moveSlots.slice(0, 4);
 			this.add('message', `${pokemon.name}'s fake claim was uncovered!`);
-			// @ts-ignore Hack for Snaquaza's Z move
-			delete pokemon.claimHP;
+			delete pokemon.m.claimHP;
```
Comment (Zarel) [inline mods/ssb/statuses.js]: Can you set it to `null`? `delete` is a mostly-unnecessary operator

**Comment by mathfreak231** [comment]
Whoops, seems we all abandoned this, I'm closing for the time being, maybe someone can come back to this?

**Comment by Zarel** [comment]
No, this is still something I'm interested in.

**Comment by Zarel** [comment]
It's just gotten lost in a bunch of other changes.

File: dev-tools/globals.ts [hunk 2]
```diff
@@ -1006,3 +1006,10 @@ interface RandomTeamsTypes {
 		fillerMoves4?: string[]
 	}
 }
+
+interface PokemonModData {
+	gluttonyFlag?: boolean; // Gen-NEXT
+	innate?: string; // Gen-NEXT
```
Comment (KrisXV) [inline dev-tools/globals.ts]: this is used for Partners in Crime, not Gen-NEXT

Comment (Slayer95) [reply dev-tools/globals.ts]: Thanks, fixed!

**Comment by Slayer95** [comment]
Rebased, @Zarel

**Comment by Slayer95** [comment]
Follow-up questions: Do we want ``m`` as a property for ``Move``? ``Side``? ``Battle``?

File: dev-tools/globals.ts [hunk 3]
```diff
@@ -1006,3 +1006,10 @@ interface RandomTeamsTypes {
 		fillerMoves4?: string[]
 	}
 }
+
+interface PokemonModData {
+	gluttonyFlag?: boolean; // Gen-NEXT
+	innate?: string; // Partners in Crime
+	originalSpecies?: string; // Mix and Mega
+	[key: string]: any;
```
Comment (Zarel) [inline dev-tools/globals.ts]: Maybe this should be `unknown` rather than `any`.

Comment (Zarel) [reply dev-tools/globals.ts]: I guess it works fine as-is.

**Comment by mathfreak231** [comment]
thanks all

**Pull request merged**